tools/codegen/dest/lazy_ir.py

```diff
@@ -168,44 +168,47 @@ class {schema.node_name} : public {self.node_base} {{
 
 
 @dataclass(frozen=True)
-class TSLazyIR(LazyIR):
+class GenTSLazyIR(GenLazyIR):
 
     def lowering_function(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> str:
         return f"""torch::lazy::TSOpVector Lower(std::shared_ptr<torch::jit::GraphFunction> function,
     torch::lazy::TSLoweringContext* loctx) const override {{
     {ts_lowering_body(f)}
   }}"""
 
-def lazy_tensor_decls(value_args: List[LazyArgument], tensor_class: str) -> str:
-    lazy_tensor_decls: List[str] = []
-    for arg in value_args:
-        if arg.is_wrapped_scalar:
-            # no lazy tensor wrapper for scalars that are promoted to IR values
-            continue
-        elif isinstance(arg.lazy_type, BaseCType):
-            if arg.lazy_type.type is tensorListValueT:
-                lazy_tensor_decls.append(
-                    f"auto lazy_{arg.name}_tensorlist = torch::lazy::GetTensorList({arg.name});")
-            else:
-                lazy_tensor_decls.append(
-                    f"{tensor_class}Ptr lazy_{arg.name} = "
-                    f"torch::lazy::GetLtcTensorOrCreateForWrappedNumber({arg.name}, *common_device);")
-        elif isinstance(arg.lazy_type, OptionalCType):
-            # TODO(alanwaketan): Maybe we want to apply GetLtcTensorOrCreateForWrappedNumber here, but hold it
-            # until we encounter a real world example.
-            lazy_tensor_decls.append(
-                f"    {tensor_class}Ptr lazy_{arg.name} = torch::lazy::TryGetLtcTensor({arg.name}.value_or(at::Tensor()));")
-        else:
-            raise AssertionError(f"TODO not sure if there are other valid types to handle here ({arg.lazy_type})")
-    return ("\n        ").join(lazy_tensor_decls)
+
 
 @dataclass(frozen=True)
-class GenLazyNativeFuncDefinition:
+class GenLazyNativeFuncDefinition(ABC):
     class_method_name: str
     backend_index: BackendIndex
     tensor_class: str
     gen_forced_fallback_code: bool
 
+    def lazy_tensor_decls(self, value_args: List[LazyArgument]) -> str:
+        # Generates lazy_{name} variables for LazyTensors wrapping input tensors
+        lazy_tensor_decls: List[str] = []
+        for arg in value_args:
+            if arg.is_wrapped_scalar:
+                # no lazy tensor wrapper for scalars that are promoted to IR values
+                continue
+            elif isinstance(arg.lazy_type, BaseCType):
+                if arg.lazy_type.type is tensorListValueT:
+                    lazy_tensor_decls.append(
+                        f"auto lazy_{arg.name}_tensorlist = torch::lazy::GetTensorList({arg.name});")
```

Comment:
@bdhirsh so in particular, there are a bunch of functions calls here that XLA wants to modify (GetTensorList, GetLtcTensorOrCreateForWrappedNumber, TryGetLtcTensor, ...)

It is possible this is just a short term need, as XLA could migrate to fully using these core APIs but wants to start off using their versions first.

We could also solve this particular problem by making something like a 'TensorFactory' interface and having Xla override that, then making codegen code just call the factory API.  This is my preferred solution in this case, but i thought it would be easier in the short term to unblock XLA using the flexible-codegen approach.

Reply:
Ah yeah - so at least in this case, it sounds like the extra extensibility is really just a temporary thing for migration, and not a long term difference between the TS/XLA backends.

TBH, if XLA doesn't want to use the existing `torch::lazy::*` API's yet, I think your `TensorFactory` interface idea sounds like it might be less work compared to building the overridable codegen, and coordinating unwinding it later on (that probably depends on how many / what sort of functions are being extended though). If you think the divergence just a temporary thing though, I think either option sounds ok.

Reply:
yea, i might be convinced in this case.

I think i still want to land this PR which actually does not make anything extensible but just refactors a bit.  I think i like the idea of making GenLazyNativeFunctionDef have a number of smaller methods and then having the __call__ just drop the result of each method into its place in the template.

Reply:
@bdhirsh wdyt about landing this as is (with a warning I added)?

I talked to @JackCaoG and it seems like there is a chicken-egg problem bc they haven't adopted LazyTensor class yet, which makes it harder for them to use our longer term API solution, but easier to get started by overriding the codegen.

Reply:
@wconstab `GenLazyNativeFuncDefinition` has a bunch stuff on it, and making the whole class abstract makes it kinda hard to reason about what functions a backend might be overriding or not (technically, someone else could override `GenLazyNativeFuncDefinition.__call__` which would be pretty weird).

What do you think of taking the set of things that XLA should be able to override, and moving them into a different extendible class somewhere (that gets called by this class)? That way XLA won't need to (or have the freedom to) override the entire `lazy_tensor_decls()` function, which is probably too much extensibility anyway. They'd have an individual method for e.g. `BackendExtensions.getTensorListAPI()`